### DISCUSSION TITLE
[draw API] draw_data vs. user_data

### DISCUSSION THREAD

**Discussion by justvanrossum** [harfbuzz/harfbuzz#3765] [Q&A] [answered]
Reading the documentation, it is not quite clear to me how the `draw_data` argument in `hb_font_get_glyph_shape()` function relates to the `user_data` argument in the `hb_draw_funcs_set_*_func()` functions.

One guess could be that the callbacks get passed `draw_data` in their `user_data` argument, if `NULL` was passed as `user_data` to `hb_draw_funcs_set_*_func()`.

**Comment by khaledhosny** [comment, marked as answer]
`user_data` is passed to the `_set_*_func` so they are fixed for a given draw function, while `draw_data` is passed to `hb_font_get_glyph_shape()` so it can be different each time this function is called (e.g. you can set the draw functions once, and re-use them for different glyphs, passing a different `draw_data` pointer for each invocation). Both pointers get passed  to the callback functions independently (the callback functions has both `draw_data` and `user_data` arguments).

**Reply by justvanrossum** [reply]
Ah yes, I somehow missed the callbacks *also* get `draw_data` passed. Thank you.
